### PR TITLE
Double Evaluation Fix

### PR DESCRIPTION
**Pull request by AndyGerlicher**
Replaces #2472 and #2458

Still WIP. I need to get the NuGet task pulled down for bootstrap build.

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
@davkean FYI

I've got some initial benchmark results for this, and it's a bit mixed. Evaluation time and memory shouldn't really be impacted. The diff here I would contribute mostly to noise. The memory seems to be about 15k more. Makes sense that with a tiny standard deviation on the small projects (GC called right before so likely no GC during test) we would see a delta. It's more to parse, few more objects materialized, etc. Surprised it would be that much, but while statistically significant I don't think it's real world significant.

Build time (design time build listed) is more interesting. For smaller projects, it's clearly not faster. And I'm not really surprised. It's an extra task from a non-msbuild dll (not ngen'd here). Many builds will end up with that dll loaded anyway (from Restore, etc.) so it's not always an extra dll, but still it's not free. And this affects all builds regardless of SDK status. I'm also a little worried this will set off RPS having the extra dll.

The good news on build time is a lot of projects should see an improvement. 30%-50% in the P2P and Picasso test cases which is great. I don't have any spinning drives to test on but I assume this could be even better there when globbing is involved.

The mixed news is testing Roslyn.sln. Doing an incremental build on the solution `msbuild /v:m Roslyn.sln` went from 335 evaluations down to 165 (good), but with `/m` the time was reliably 27s -> 26s. Which is not that spectacular. Without /m the times was a little more pronounced at 1:36 -> 1:23. Which after calculating is 13% so that is pretty good, just not as much as I was hoping for. Perhaps it will be more for design-time builds, but I couldn't get that working.

### Evaluation: Time (ms)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 30.395 -> 30.5164 (0.399%)
DotnetWebProject | :red_circle: | yes | 49.611 -> 49.6851 (0.149%)
DotnetMvcProject | :red_circle: | yes | 86.4774 -> 86.5475 (0.081%)
Picasso | :white_check_mark: | yes | 266.2479 -> 256.1184 (-3.805%)
SmallP2POldCsproj | :white_check_mark: | yes | 42.6504 -> 41.8708 (-1.828%)
SmallP2PNewCsproj | :red_circle: | yes | 82.5688 -> 82.6155 (0.057%)
Generated_100_100_v150 | :red_circle: | yes | 1254.7972 -> 1290.9703 (2.883%)
LargeP2POldCsproj | :white_check_mark: | yes | 753.0709 -> 751.137 (-0.257%)
roslyn | :white_check_mark: | yes | 5370.7354 -> 5334.5412 (-0.674%)

### Evaluation: Memory (bytes)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 4944663 -> 4959678 (0.304%)
DotnetWebProject | :red_circle: | yes | 6840943 -> 6861401 (0.299%)
DotnetMvcProject | :red_circle: | yes | 9028362 -> 9043496 (0.168%)
Picasso | :red_circle: | yes | 35430020 -> 35436938 (0.02%)
SmallP2POldCsproj | :white_check_mark: | yes | 6034210 -> 5993157 (-0.68%)
SmallP2PNewCsproj | :red_circle: | yes | 10577958 -> 10602931 (0.236%)
Generated_100_100_v150 | :red_circle: | yes | 147579824 -> 147587200 (0.005%)
LargeP2POldCsproj | :red_circle: | yes | 86536590 -> 86568520 (0.037%)
roslyn | :red_circle: | yes | 585084467 -> 585118839 (0.006%)

### Build: Time (ms)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 7.5323 -> 7.6717 (1.851%)
DotnetWebProject | :red_circle: | yes | 9.2563 -> 9.39 (1.444%)
DotnetMvcProject | :red_circle: | yes | 10.0831 -> 10.2063 (1.222%)
Picasso | :white_check_mark: | yes | 2382.0946 -> 1020.6393 (-57.154%)
SmallP2POldCsproj | :white_check_mark: | yes | 84.8296 -> 48.2654 (-43.103%)
SmallP2PNewCsproj | :white_check_mark: | yes | 87.373 -> 53.141 (-39.179%)

### Build: Memory (bytes)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 1151919 -> 1174937 (1.998%)
DotnetWebProject | :white_check_mark: | yes | 1302788 -> 1302090 (-0.054%)
DotnetMvcProject | :red_circle: | yes | 1476974 -> 1482962 (0.405%)
Picasso | :white_check_mark: | yes | 277497653 -> 125792697 (-54.669%)
SmallP2POldCsproj | :white_check_mark: | yes | 9756337 -> 6211117 (-36.338%)
SmallP2PNewCsproj | :white_check_mark: | yes | 10009262 -> 6611995 (-33.941%)

**Comment by AndyGerlicher** [comment]
FYI this will be CI red until https://github.com/NuGet/NuGet.Client/pull/1737 goes in and publishes and I can update the package version. In #2594 I updated our bootstrap to pull that task from a package so it should be just a version bump.

If anyone wants to set this up locally, run `cibuild --boostrap-only` and edit NuGet.targets to match the one in the NuGet PR. Then add a `Directory.Build.props` with this (replacing the path to where you built NuGet):
```XML
<Project>
<PropertyGroup>
 <RestoreTaskAssemblyFile>D:\src\NuGet.Client\artifacts\NuGet.Build.Tasks\15.0\bin\Debug\net45\NuGet.Build.Tasks.dll</RestoreTaskAssemblyFile>
</PropertyGroup>
</Project>
```

File: src/Tasks/Microsoft.Common.CrossTargeting.targets [hunk 1]
```diff
@@ -20,6 +20,19 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
 
+  <Target Name="GetTargetFrameworks"
+          Returns="@(_ThisProjectBuildMetadata)">
+    <ItemGroup>
+      <_ThisProjectBuildMetadata Include="$(MSBuildProjectFullPath)">
+        <TargetFrameworks Condition="'$(TargetFrameworks)' != ''">$(TargetFrameworks)</TargetFrameworks>
+        <TargetFrameworks Condition="'$(TargetFrameworks)' == ''">$(TargetFramework)</TargetFrameworks>
+        <HasSingleTargetFramework>true</HasSingleTargetFramework>
+        <HasSingleTargetFramework Condition="'$(IsCrossTargetingBuild)' == 'true'">false</HasSingleTargetFramework>
+        <IsRidAgnostic>true</IsRidAgnostic>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.Common.CrossTargeting.targets]: Why is this hard-coded?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -1558,6 +1544,61 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+
+  <Target Name="GetTargetFrameworks"
+          Returns="@(_ThisProjectBuildMetadata)">
+    <ItemGroup>
+      <_ThisProjectBuildMetadata Include="$(MSBuildProjectFullPath)">
+        <TargetFrameworks Condition="'$(TargetFrameworks)' != ''">$(TargetFrameworks)</TargetFrameworks>
+        <TargetFrameworks Condition="'$(TargetFrameworks)' == ''">$(TargetFramework)</TargetFrameworks>
+        <HasSingleTargetFramework>true</HasSingleTargetFramework>
+        <HasSingleTargetFramework Condition="'$(IsCrossTargetingBuild)' == 'true'">false</HasSingleTargetFramework>
+        <IsRidAgnostic>true</IsRidAgnostic>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: Ditto

File: src/Tasks/Microsoft.Common.CrossTargeting.targets (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.Common.CrossTargeting.targets]: See https://github.com/dotnet/sdk/blob/d7730b73a5bea6c1e7f5d178711322082c93d96e/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets#L71-L73

and https://github.com/dotnet/sdk/issues/696, which might regress.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 3]
```diff
@@ -1558,6 +1544,61 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+  </Target>
+
+  <Target Name="GetTargetFrameworks"
+          Returns="@(_ThisProjectBuildMetadata)">
+    <ItemGroup>
+      <_ThisProjectBuildMetadata Include="$(MSBuildProjectFullPath)">
+        <TargetFrameworks Condition="'$(TargetFrameworks)' != ''">$(TargetFrameworks)</TargetFrameworks>
+        <TargetFrameworks Condition="'$(TargetFrameworks)' == ''">$(TargetFramework)</TargetFrameworks>
+        <HasSingleTargetFramework>true</HasSingleTargetFramework>
+        <HasSingleTargetFramework Condition="'$(IsCrossTargetingBuild)' == 'true'">false</HasSingleTargetFramework>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: I think this is always going to be false when we land in the non-cross-targeting version of GetTargetFrameworks.

File: src/Tasks/Microsoft.Common.CrossTargeting.targets (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.Common.CrossTargeting.targets]: We need to get the SDK tests running with this change. cc @dsplaisted

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -1558,13 +1569,25 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     </MSBuild>
 
     <!-- For each reference, get closest match -->
-    <AssignReferencePropertiesTask AnnotatedProjectReferences="@(_ProjectReferenceTargetFrameworkPossibilities)"
-                                   CurrentProjectTargetFramework="$(TargetFrameworkMoniker)"
-                                   Condition="'@(_ProjectReferenceTargetFrameworkPossibilities->Count())' != '0'">
+    <GetReferenceNearestTargetFrameworkTask AnnotatedProjectReferences="@(_ProjectReferenceTargetFrameworkPossibilities)"
+                                            CurrentProjectTargetFramework="$(ReferringTargetFrameworkForProjectReferences)"
+                                            CurrentProjectName="MSBuildProjectName"
```
Comment (nguerrera) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: $(MSBuildProjectName) ?

**Comment by AndyGerlicher** [comment]
Updated numbers:

### Evaluation: Time (ms)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 29.0593 -> 28.9913 (-0.234%)
DotnetWebProject | :white_check_mark: | yes | 44.2305 -> 44.1433 (-0.197%)
DotnetMvcProject | :red_circle: | yes | 50.4046 -> 50.653 (0.493%)
Picasso | :white_check_mark: | yes | 256.6368 -> 255.6237 (-0.395%)
SmallP2POldCsproj | :red_circle: | yes | 42.4037 -> 42.7942 (0.921%)
SmallP2PNewCsproj | :red_circle: | yes | 77.9943 -> 78.3615 (0.471%)
Generated_100_100_v150 | :white_check_mark: | yes | 1260.2155 -> 1257.19 (-0.24%)
LargeP2POldCsproj | :red_circle: | yes | 742.7307 -> 744.3613 (0.22%)
roslyn | :white_check_mark: | yes | 3306.3346 -> 3275.4293 (-0.935%)

### Evaluation: Memory (bytes)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 4895467 -> 4893821 (-0.034%)
DotnetWebProject | :red_circle: | yes | 6561947 -> 6565272 (0.051%)
DotnetMvcProject | :red_circle: | yes | 7408101 -> 7414780 (0.09%)
Picasso | :white_check_mark: | yes | 35314766 -> 35290989 (-0.067%)
SmallP2POldCsproj | :white_check_mark: | yes | 6017226 -> 5984496 (-0.544%)
SmallP2PNewCsproj | :red_circle: | yes | 10358236 -> 10366868 (0.083%)
Generated_100_100_v150 | :red_circle: | yes | 147536691 -> 147553504 (0.011%)
LargeP2POldCsproj | :red_circle: | yes | 86257330 -> 86264699 (0.009%)
roslyn | :ok_hand: | no | 407241006 -> 408094764 (0.21%)

### Build: Time (ms)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 7.597 -> 7.563 (-0.448%)
DotnetWebProject | :white_check_mark: | yes | 9.2922 -> 9.2719 (-0.218%)
DotnetMvcProject | :red_circle: | yes | 9.9721 -> 10.0969 (1.251%)
Picasso | :white_check_mark: | yes | 2309.7216 -> 1015.8582 (-56.018%)
SmallP2POldCsproj | :white_check_mark: | yes | 84.6579 -> 48.6706 (-42.509%)
SmallP2PNewCsproj | :white_check_mark: | yes | 82.0603 -> 50.1221 (-38.92%)

### Build: Memory (bytes)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 1155168 -> 1154402 (-0.066%)
DotnetWebProject | :red_circle: | yes | 1279572 -> 1281203 (0.127%)
DotnetMvcProject | :red_circle: | yes | 1452351 -> 1457735 (0.371%)
Picasso | :white_check_mark: | yes | 275855704 -> 125280131 (-54.585%)
SmallP2POldCsproj | :white_check_mark: | yes | 9746673 -> 6192245 (-36.468%)
SmallP2PNewCsproj | :white_check_mark: | yes | 9733262 -> 6400177 (-34.244%)

File: documentation/ProjectReference-Protocol.md [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿# The `ProjectReference` Protocol
+# The `ProjectReference` Protocol
```
Comment (dsplaisted) [inline documentation/ProjectReference-Protocol.md]: Are the changes to this file up-to-date after the changes to the protocol and interface with NuGet that were made in this PR?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 6]
```diff
@@ -1516,17 +1516,26 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
 
-    Builds the GetTargetFrameworkProperties target of all existent project references,
-    passing $(TargetFrameworkMoniker) as $(ReferringTargetFramework) and sets the
-    SetTargetFramework metadata of the project reference to the value that is returned.
-
-    This allows a cross-targeting project to select how it should be configured to
-    build against the most appropriate target for the referring target framework.
+    Builds the GetTargetFrameworks target of all existent project references to get a list
+    of all supported TargetFrameworks of the referenced projects. For each project, call
+    GetReferenceNearestTargetFrameworkTask to determine the closest match. This allows a
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: Reword this to say "Call GetReferenceNearestTargetFrameworkTask to determine the closest match for each project."  This way it won't imply that the task is called once for project.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 7]
```diff
@@ -1558,50 +1567,64 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+                                            Condition="'@(_ProjectReferenceTargetFrameworkPossibilities->Count())' != '0' and '$(ReferringTargetFrameworkForProjectReferences)' != ''">
+      <Output ItemName="AnnotatedProjects" TaskParameter="AssignedProjects" />
+    </GetReferenceNearestTargetFrameworkTask>
 
-        <UndefineProperties Condition="$(_ProjectReferenceTargetFrameworkProperties.Contains(`ProjectHasSingleTargetFramework=true`))">%(_MSBuildProjectReferenceExistent.UndefineProperties);TargetFramework;ProjectHasSingleTargetFramework</UndefineProperties>
-        <!-- Unconditionally remove the property that was set as a marker to indicate that for this call we should remove TargetFramework -->
-        <UndefineProperties Condition="!$(_ProjectReferenceTargetFrameworkProperties.Contains(`ProjectHasSingleTargetFramework=true`))">%(_MSBuildProjectReferenceExistent.UndefineProperties);ProjectHasSingleTargetFramework</UndefineProperties>
-      </_MSBuildProjectReferenceExistent>
+    <ItemGroup>
+      <AnnotatedProjects Include="@(_ProjectReferenceTargetFrameworkPossibilities)"
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: I'd suggest adding a comment here explaining this logic (possibly linking to https://github.com/dotnet/sdk/issues/416)

**Comment by AndyGerlicher** [comment]
.NET Core doesn't care if the dlls are delay-signed, but we're not going to get a green build for full framework until it's signed.
